### ISSUE TITLE
Pyright does not deduce types that it should

### ISSUE THREAD

**Issue by acampove** [microsoft/pyright#11530]
**Describe the bug**

There seem to be cases where the type is obvious, but pyright needs to be explicitly told. This potentially could hide bugs in the user's code.

**Code or Screenshots**

```python
class Foo:
    bar: list[int]

foo = Foo()

components = foo.bar
components = 3
```

I see:

```
(rx_run3) [acampove@thinkbook explicit-annotation]$ pyright test.py
0 errors, 0 warnings, 0 informations
(rx_run3) [acampove@thinkbook explicit-annotation]$ cat pyrightconfig.json
{
  "typeCheckingMode": "strict"
}
```
Here components could be reasigned incorrectly later.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

I am using the CLI, but the same problem happens in neovim. The version is:

```
pyright 1.1.411
```

**Comment by pinterior** [comment]
https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-declarations
> If a variable or parameter has no type annotation, Pyright will assume that any value can be assigned to it.


after the first assignment, `component` has type `list[int]`; after second assignment, it has type `int`.

**Comment by rchiodo** [comment]
As @pinterior pointed out. If you don't give a variable a type, Pyright assumes its intended type is Any and allows all assignments to it. Even in strict mode. 

I suppose this might be considered an enhancement request to catch these situations? I think that's likely outside the scope of Pyright though and more something like what Ruff would do. 

Pyright checks for things that will cause runtime errors but doesn't check for things that are generally considered linting problems.

**Comment by acampove** [comment]
> If a variable or parameter has no type annotation, Pyright will assume that any value can be assigned to it

But pyright knows the type of the value that `component` is assigned to, so the user should not have to specify it by hand. That is just bad design and will make the user's code less readable. @pinterior @rchiodo

**Comment by rchiodo** [comment]
This may be your intent though (a variable that changes types). If you want a variable to always have a specific type, then you'd give it a type annotation. Pyright will check that.
